### DISCUSSION TITLE
Example x-chat.php

### DISCUSSION THREAD

**Discussion by carliedu** [clue/framework-x#230] [Q&A] [answered]
Hi,
as suggested by @simonfrings, I'm creating this discussion.
I'm trying to use Framework-X the same way I used ReactPHP for an project that interfaces real-time with users.
The same way as I started the ReactPHP project, the first simple application that must work is a chat.
Chat is an application that interfaces with user on a Browser, and that is, at least for me, an typical internet use where user sends and receives data real-time, asynchronously.

For implement this with Framework-X, I turned to the example chat that @clue created in the example project clue/php-sse-react. This works fine, despite being quite simple.

First I created a fork from ```examples/index.php``` of Framework-X as ```x-chat.php```.
I'm using the same files ```10-eventsource.html``` and ```10-styles.css``` necessary as interface.

Opening ```11-chat.php``` and ```x-caht.php``` at the same time, I adapted the code in ```x-chat.php``` to do the same as ```11-chat.php```. I added code to get the PORT number that I desire to be used, different from default 8080 from Framework-X, and included some ```echo``` commands to see on console, where the the code is passing. I needed to add ```use React\Stream\ThroughStream;``` and ```$stream = new ThroughStream();``` (and it works). But I don't know how to implement ```$loop->futureTick(function () use ($channel, $stream, $id)```.

I am aware that the code will give error, but I tryied.
Runnig the code with:
```php x-chat.php 9999```
and opening a browser and calling ```http://myserver.com:9999/```, I'm getting following on console:
```
24/06/2023-12:04:49 (/x-chat.php) *---INI---INI---INI---INI---INI--
2023-06-24 12:04:49.713 Listening on http://0.0.0.0:53041
24/06/2023-12:04:56 (/x-chat.php) getUri /   -> Send 10-eventsource.html
2023-06-24 12:04:56.188 191.204.182.47 "GET / HTTP/1.1" 200 3196 0.003

Warning: Undefined variable $loop in /usr/local/src/github/carliedu/framework-x/examples/x-chat/x-chat.php on line 64
2023-06-24 12:04:56.382 191.204.182.47 "GET /chat HTTP/1.1" 500 1294 0.001
24/06/2023-12:04:56 (/x-chat.php) getUri /styles.css   -> Send 10-styles.css
2023-06-24 12:04:56.532 191.204.182.47 "GET /styles.css HTTP/1.1" 200 2909 0.001
```
Looking Framework-X documentation, I don't know what to do with $loop that is so obvious on ReactPHP, but not here.

I noted that the file 10-styles.css seems to be loaded (last line of console), but do not reflect on browser.

For now I stopped here. I need help.

Best,
carliedu

**Comment by bpolaszek** [comment]
Can you please provide a repository with your example code? Hard to assist you with debugging otherwise.

**Reply by carliedu** [reply]
Excuse me @bpolaszek. I forgot to mention it. 
https://github.com/carliedu/framework-x/tree/main/examples/x-chat

**Reply by bpolaszek** [reply, marked as answer]
Alright, I think you forgot [this](https://github.com/clue/php-sse-react/blob/master/examples/11-chat.php#L10) on top of your script (which is actually the old way of instantiating the loop - you can skip this, and line 64, replace `$loop` by `Loop::get()` as it's been replaced by a singleton).

**Reply by carliedu** [reply]
@bpolaszek, thanks for responding.
But my doubts are o lot greater.
Changing ```$loop->futureTick``` by ```Loop::get``` makes no sense for me. I think it could be ```Loop::futureTick```.
But all the logic for ```/chat``` is strange for me, beginning with the object ```$channel``` from class ```BufferedChannel```.

Another thing I could not solve is sending ```10-styles.css``` as as type ```text/css```. I tried it with ´´´text/html``` without success.

Best,

**Reply by SimonFrings** [reply]
I agree with @bpolaszek! After taking a look at your `x-chat.php` I recognized that you never assigned anything to the `$loop` variable. This is also why you get the warning: `Undefined variable $loop in...`

> Changing $loop->futureTick by Loop::get makes no sense for me. I think it could be Loop::futureTick.
But all the logic for /chat is strange for me, beginning with the object $channel from class BufferedChannel.

I think what @bpolaszek tried to say is that you need to get the currently active event loop instance before trying to call the `futureTick` method:

```php
…

$loop = Loop::get();

$loop->futureTick(function () use ($channel, $stream, $id) {
    $channel->connect($stream, $id);
});

…
```

You can read more about `Loop::get()` here: https://github.com/reactphp/event-loop#get

> Another thing I could not solve is sending 10-styles.css as as type text/css. I tried it with ´´´text/html``` without success.

You can either use the response methods from https://github.com/reactphp/http like you're currently doing in `x-chat.php` or you can create a response like this:

```php
$html = <<<HTML
<!doctype html>
<html>
<body>Hello wörld!</body>
</html>

HTML;

$response = new React\Http\Message\Response(
    React\Http\Message\Response::STATUS_OK,
    [
        'Content-Type' => 'text/html; charset=utf-8'
    ],
    $html
);
```

For further information take a look at: https://github.com/reactphp/http#response

Hope this helps :+1:

**Comment by carliedu** [comment]
Hi @SimonFrings and @bpolaszek.
Thanks for your support.
I'm a pretty lost with the sintax used in Framework-X.
Two years ago, as I discovered ReactPHP, it was the same. Until I finally created my first program, it toke me months to understand how to build it. The final result was fantastic, as I managed to communicate asynchronously with an exchange and with several users in browsers receiving information asynchronously as it happened.
To reach the result I had a lot of difficulty, mainly in contacting the client in the browser. I was only able to take the first step when I found the example 11-chat.php from @clue.
For me, I ended up making a parallel with the circular logic of RPG (which I used at the beginning of my career) and Easytrieve, both in Mainframe.
So, all that is between ```$loop = React\EventLoop\Factory::create();``` and ```$loop->run();``` runs in a loop, until the logic stops it.
Now, with the ```index.php``` example from Framework-X, I'm not identifying this loop. So I don't know what must be coded before and after the LOOP, and what must be code in the LOOP.
So the classes ```BufferedChannel``` and ```ThroughStream``` I don't know where to put them (and I must sincerely say, don't understand how tey work - I know they work in the ReactPHP logic).

For now, I'm supposing that in Framework-X, the loop begins with ```$app = new FrameworkX\App($container);``` and ends with ```$app->run();```.
The ReactPHP ```switch/case``` is treated in Framework-X as ```$app->get('XXXXXXX```.
The ```$app->get('/'``` and ```$app->get('/style'``` are working fine (as I made the correction cited by @SimonFrings.
The error occurs in the logic of ```$app->get('/chat'``` where ```ThrughStream```, ```futureTick``` and ```BufferedChannel``` must be implemented.

I believe that all my problems are related to BufferedChannel. I don't understand exactly what and how it works, and consequently don't know where and how it must be coded.

Any help would be appreciated.

Best,
carliedu
